### PR TITLE
[db] ResetSnapshots() performance improvement

### PR DESCRIPTION
**Pull request by millken**
# Description
Reduced the computation of map keyhash, reduced `keyCacheValue` memory allocation, Mintblock time dropped by half, this PR has been tested on nightly

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [x] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: db/batch/batch_impl.go [hunk 1]
```diff
@@ -344,8 +344,9 @@ func (cb *cachedBatch) RevertSnapshot(snapshot int) error {
 	for tag := cb.tag; tag < len(cb.tagKeys); tag++ {
 		keys := cb.tagKeys[tag]
 		for _, key := range keys {
-			cb.keyTags[key] = cb.keyTags[key][:len(cb.keyTags[key])-1]
-			if len(cb.keyTags[key]) == 0 {
+			kv := cb.keyTags[key]
+			kv.pop()
```
Comment (CoderZhi) [inline db/batch/batch_impl.go:348]: reverse

**Review by CoderZhi — COMMENTED** [review]
any benchmark test?

**Comment by millken** [comment]
> any benchmark test?

update benchmark

File: db/batch/batch_impl.go [hunk 2]
```diff
@@ -344,8 +344,9 @@ func (cb *cachedBatch) RevertSnapshot(snapshot int) error {
 	for tag := cb.tag; tag < len(cb.tagKeys); tag++ {
 		keys := cb.tagKeys[tag]
 		for _, key := range keys {
-			cb.keyTags[key] = cb.keyTags[key][:len(cb.keyTags[key])-1]
-			if len(cb.keyTags[key]) == 0 {
+			kv := cb.keyTags[key]
+			kv.reverse()
```
Comment (CoderZhi) [inline db/batch/batch_impl.go:348]: the function name should be `pop`.

**Review by CoderZhi — APPROVED** [review]

**Review by envestcc — APPROVED** [review]

File: db/batch/kv_cache.go [hunk 3]
```diff
@@ -39,6 +40,38 @@ type (
 	}
 )
 
+func newkvCacheValue(v []int) *kvCacheValue {
+	return (*kvCacheValue)(&v)
+}
+
+func (kv *kvCacheValue) reset() {
+	([]int)(*kv)[0] = 0
+	*kv = (*kv)[0:1]
```
Comment (dustinxie) [inline db/batch/kv_cache.go:49]: `[:1]`

File: db/batch/kv_cache.go [hunk 4]
```diff
@@ -39,6 +40,38 @@ type (
...
+}
+
+func (kv *kvCacheValue) reverse() {
+	*kv = (*kv)[0 : kv.len()-1]
```
Comment (dustinxie) [inline db/batch/kv_cache.go:53]: `[:kv.len()-1]` and name this func `pop`

**Pull request merged**